### PR TITLE
varlink: add a couple of tests + accompanying fixes

### PR DESCRIPTION
**Pull request by mrc0mmand**
Add a couple of tests for varlinkctl, a fuzzer for the varlink IDL stuff, and fix a couple of things along the way.

### CODE REVIEW

File: src/fuzz/fuzz-varlink-idl.c [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+#include "varlink-idl.h"
+#include "log.h"
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+        _cleanup_(varlink_interface_freep) VarlinkInterface *vi = NULL;
+        _cleanup_free_ char *str = NULL, *dump = NULL;
+        int r;
+
+        if (!getenv("SYSTEMD_LOG_LEVEL"))
+                log_set_max_level(LOG_CRIT);
```
Comment (poettering) [inline src/fuzz/fuzz-varlink-idl.c]: why not do:

```c
log_set_max_level(LOG_CRIT);
log_parse_environment();
(void) log_open();
```
that way you first set the log level to LOG_CRIT, but then you allow the env var to override it

File: src/shared/varlink-idl.c [hunk 2]
```diff
@@ -1004,6 +1004,9 @@ int varlink_idl_parse(
                 case STATE_METHOD_ARROW:
                         assert(symbol);
 
+                        if (!token)
+                                return log_debug_errno(SYNTHETIC_ERRNO(EBADMSG), "%u:%u: Premature EOF.", *line, *column);
```
Comment (poettering) [inline src/shared/varlink-idl.c:1007]: can you add this to the STATE_METHOD thing above too? and actually STATE_TYPE, STATE_ERROR below too.

seems i forgot that not just once but four times

**Comment by poettering** [comment]
looks excellent! thank you so much for this!

**Comment by keszybz** [comment]
2536s   Temporary failure resolving 'ftpmaster.internal'
2536s E: Failed to fetch http://ftpmaster.internal/ubuntu/pool/universe/n/netlabel-tools/netlabel-tools_0.30.0-1_amd64.deb  Temporary failure resolving 'ftpmaster.internal'

It's Friday and the FTP Master has gone AWOL.

**Comment by keszybz** [comment]
Yeah, it seems pcrextend is not built for ppc64el and s390x. Let's see if it passes now.

**Comment by bluca** [comment]
> 2536s Temporary failure resolving 'ftpmaster.internal' 2536s E: Failed to fetch http://ftpmaster.internal/ubuntu/pool/universe/n/netlabel-tools/netlabel-tools_0.30.0-1_amd64.deb Temporary failure resolving 'ftpmaster.internal'
> 
> It's Friday and the FTP Master has gone AWOL.

No we broke that, see: https://github.com/systemd/systemd/pull/29552

**Comment by mrc0mmand** [comment]
> 2536s Temporary failure resolving 'ftpmaster.internal' 2536s E: Failed to fetch http://ftpmaster.internal/ubuntu/pool/universe/n/netlabel-tools/netlabel-tools_0.30.0-1_amd64.deb Temporary failure resolving 'ftpmaster.internal'
> 
> It's Friday and the FTP Master has gone AWOL.

This is actually another issue - https://github.com/systemd/systemd/pull/29552 - that still needs some investigation.

**Review by bluca — APPROVED** [review]

**Comment by keszybz** [comment]
5303s testsuite-74.sh[1016]: + systemctl start systemd-userdbd
5305s [  836.042461] systemctl[1039]: Failed to start systemd-userdbd.service: Unit systemd-userdbd.service not found.

So it seems this also needs to be conditionalized somehow.

File: test/README.testsuite [hunk 3]
```diff
@@ -305,3 +307,28 @@ in the test definition (i.e. TEST-*-NAME/test.sh), which will skip the post-test
...
+
+Running fuzzers locally
+=======================
+
+When introducing a new fuzzer you usually want to check how well it behaves, as
+well as to check if it triggers anything unusual straight away. The easiest
+way is to use the LLVM's libFuzzer engine [0] together with ASan and UBSan:
+
+$ CC=clang CXX=clang++ meson setup build-libfuzz -Dllvm-fuzz=true -Db_sanitize=address,undefined -Db_lundef=false
+$ ninja -C build-libfuzz
```
Comment (evverx) [inline test/README.testsuite]: `ninja -C build-libfuzz fuzzers` should speed it up a bit.

I think
```sh
SANITIZER=address,undefined ./tools/oss-fuzz.sh
```
was supposed to be used to build the fuzz targets locally. It also sets `FUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION` as far as I can remember.

Comment (evverx) [reply test/README.testsuite]: I'd also mention `UBSAN_OPTIONS=print_stacktrace=1:print_summary=1:halt_on_error=1` somewhere. Without that the fuzz targets don't fail under UBSan.

File: src/fuzz/fuzz-varlink-idl.c (hunk 1 above)
Comment (evverx) [reply src/fuzz/fuzz-varlink-idl.c]: I think it should set the log target to `LOG_TARGET_NULL` too. The fuzz targets shouldn't log anything on OSS-Fuzz (or in any other fuzzing environment) . That's probably something that should be fixed in all the fuzz targets though.

File: test/README.testsuite (hunk 3 above)
Comment (mrc0mmand) [reply test/README.testsuite]: I knew I saw the `oss-fuzz.sh` stuff somewhere, and it's, of course, already documented in `docs/HACKING.md`, so I updated it with the stuff from this commit (and hopefully addressed the comments as well).

**Review by DaanDeMeyer — APPROVED** [review]

**Pull request merged**